packages/core/src/routes/session/utils.test.ts

```diff
@@ -0,0 +1,179 @@
+  addOIDCScope: typeof grantAddOIDCScope;
+  addResourceScope: typeof grantAddResourceScope;
+
+  constructor() {
+    this.save = grantSave;
+    this.addOIDCScope = grantAddOIDCScope;
+    this.addResourceScope = grantAddResourceScope;
+  }
+}
+
+const createContext = () => ({
+  ...createMockContext(),
+  addLogContext: jest.fn(),
+  log: jest.fn(),
+});
+
+const createProvider = () => new Provider('');
+
+jest.mock('oidc-provider', () => ({
+  Provider: jest.fn(() => ({
+    Grant,
+    interactionDetails,
+    interactionResult,
+  })),
+}));
+
+afterEach(() => {
+  grantSave.mockClear();
+  interactionResult.mockClear();
+});
+
+describe('signInWithPassword()', () => {
+  it('assign result', async () => {
+    interactionDetails.mockResolvedValueOnce({ params: {} });
+    await signInWithPassword(createContext(), createProvider(), {
+      identifier: SignInIdentifier.Username,
+      password: 'password',
+      findUser: jest.fn(async () => mockUser),
+      logType: 'SignInUsernamePassword',
+      logPayload: { username: 'username' },
+    });
+    expect(interactionResult).toHaveBeenCalledWith(
+      expect.anything(),
+      expect.anything(),
+      expect.objectContaining({ login: { accountId: mockUser.id } }),
+      expect.anything()
+    );
+  });
+
+  it('throw if user not found', async () => {
+    interactionDetails.mockResolvedValueOnce({ params: {} });
+    await expect(
+      signInWithPassword(createContext(), createProvider(), {
+        identifier: SignInIdentifier.Username,
+        password: 'password',
+        findUser: jest.fn(async () => null),
+        logType: 'SignInUsernamePassword',
+        logPayload: { username: 'username' },
+      })
+    ).rejects.toThrowError();
+  });
+
+  it('throw if user found but wrong password', async () => {
+    interactionDetails.mockResolvedValueOnce({ params: {} });
+    await expect(
+      signInWithPassword(createContext(), createProvider(), {
+        identifier: SignInIdentifier.Username,
+        password: '_password',
+        findUser: jest.fn(async () => mockUser),
+        logType: 'SignInUsernamePassword',
+        logPayload: { username: 'username' },
+      })
+    ).rejects.toThrowError();
+  });
+
+  it('throw if sign in method is not enabled', async () => {
+    findDefaultSignInExperience.mockResolvedValueOnce({
+      ...mockSignInExperience,
+      signIn: {
+        methods: [
+          {
+            ...mockSignInMethod,
+            identifier: SignInIdentifier.Sms,
+            password: false,
+          },
+        ],
+      },
+    });
+    interactionDetails.mockResolvedValueOnce({ params: {} });
+    await expect(
+      signInWithPassword(createContext(), createProvider(), {
+        identifier: SignInIdentifier.Username,
+        password: 'password',
+        findUser: jest.fn(async () => mockUser),
+        logType: 'SignInUsernamePassword',
+        logPayload: { username: 'username' },
+      })
+    ).rejects.toThrowError();
```

Comment:
nit: check error by error code could be more accurate?

Reply:
updated